### ISSUE TITLE
ERROR: timeout while waiting for ocf process to start

### ISSUE THREAD

**Issue by darsto** [spdk/spdk#815]
Intermittent failure. OCF test hangs inside waitforlisten.

```
Starting SPDK v19.07-pre / DPDK 19.05.0 initialization...
[ DPDK EAL parameters: iscsi --no-shconf -c 0x1 --log-level=lib.eal:6 --log-level=lib.cryptodev:5 --log-level=user1:6 --base-virtaddr=0x200000000000 --match-allocations --file-prefix=spdk_pid3611079 ]
EAL: No available hugepages reported in hugepages-1048576kB
EAL: VFIO support initialized
app.c: 628:spdk_app_start: *NOTICE*: Total cores available: 1
reactor.c: 251:_spdk_reactor_run: *NOTICE*: Reactor started on core 0
ocfPT: Cannot find core 1 in pool, core added as inactive
ocfWT: Cannot find core 1 in pool, core added as inactive
 13:05:08	# xtrace_enable
 13:05:08	# ((  i == 0  ))
 13:05:08	# echo 'ERROR: timeout while waiting for process (pid: 3611079) to start listening on '\''/var/tmp/spdk.sock'\'''
ERROR: timeout while waiting for process (pid: 3611079) to start listening on '/var/tmp/spdk.sock'
```

https://dqtibwqq6s6ux.cloudfront.net/results/autotest-per-patch/builds/34144/archive/nvme_phy_autotest/build.log

**Comment by tomzawadzki** [comment]
Just throwing it out there. Comparing the attached log to one that passes, it seems like the ocfWB0/1 do not finish the examine. The flow should be that:
1) aio4 is examined
2) ocfWB0 starts the cache, but does not attach cores which are not present yet (aio5/aio6) - but 'adds them as inactive'
3) ocfWB1 attaches to already started cache
4) during aio5/6 examine the cores are attached
From the failing log it seems that starting from step 2), this flow is not followed.

**Comment by darsto** [comment]
@Donaim is on two-week vacation and we don't have anyone to work on this right now, so I suggest disabling the test. I pushed a patch: https://review.gerrithub.io/c/spdk/spdk/+/457483

**Comment by darsto** [comment]
The test is disabled for now.

**Comment by tomzawadzki** [comment]
Bug scrub: @Donaim working on reproducing the issue

**Comment by mdziegie** [comment]
I spent some time on debugging this issue, conclusions:

1. Bdev aio has IO in flight limit equal to 128 for all aio instances. In case when we have few aio devices (like in this case) we can easily get -EAGAIN from io_submit syscall.

2. From bdev layer perspective in above case we get SPDK_BDEV_IO_STATUS_NOMEM, and we add it to nomem_io list. Unfortunately, if there was only one in-flight IO (from this bdev perspective) we will stuck here, because:

- in _spdk_bdev_io_do_submit function, if nomem_io isn't empty we will add new IO to this list instead of send it

- to trigger _spdk_bdev_ch_retry_io we need io completion - we never receive it because we are not sending new IOs

3. During debugging few times I caught Assert (on null check) in OCF, caused by improper initialization of ocf instance:
Probably by hung mentioned above, OCF hit timeout during register_path in mngt_poll_fn. On timeout we are going to the next step and in this situation we skipped add_core step, this caused improper initialization. It was very rare case, but I think that we should improve handling timeouts in register_path.

**Comment by jimharris** [comment]
I think this bug is related to some of the bdev_aio changes made earlier this year.  We switched AIO to use a single context per thread, rather than one per channel.  As you've noticed, the IO_STATUS_NOMEM handling only works if the NOMEM condition is associated with a single channel - i.e. we are guaranteed to get a completion associated with the channel so that the next IO can be sent.

We've talked in the past (I know @darsto has especially pushed for this) about adding additional timers to kick the nomem_io list so that it's not solely dependent on subsequent completions.  There were performance/efficiency reasons driving the bdev_aio changes made earlier this year - so reverting those isn't ideal.  Probably adding the timers is the best path forward (and would cover other cases where this may come up in the future in other bdevs).  My main concern is making sure we allocate and use the timers efficiently.
